refactor: split out debug/diagnostic infra into new crate

bitwalker opened the pull request:
This PR moves the debug information and diagnostics-related components from `miden-core` and `miden-assembly-syntax` respectively, into a new crate, `miden-core-diagnostics`.

**NOTE:** This should not be merged until #1937 is merged _and_ I've been able to update this branch to use the crates.io release of `serde_spanned`.

A few notes:

* We currently have a `miden-diagnostics` crate that is now only used by AirScript. I'd really like to move our diagnostics infrastructure there, but that's a bigger project than I'm interested in tackling right now.
* This leaves `miden-core` dependent on the new `miden-core-diagnostics` crate so that we can preserve the exports of the debug info types, and the `sync` module. AFAIK, we were only really using the `sync` module in the source management implementation, so if we don't really need to continue re-exporting that from `miden-core`, nor the `debuginfo` module and its contents, then we could remove the dependency on `miden-core-diagnostics` entirely
* The way things are exported from `miden-core-diagnostics` is a bit different, now that the debug info _and_ diagnostics components are mixed together. The top-level exports are, as you'd expect, primarily diagnostics-related. If we would prefer an alternative organization, I'm open to suggestions. The main point of annoyance is how pervasively things are re-exported from different crates (and different levels within those crates). If I had my way, we'd nuke all of that and stick to more explicit imports, but in an effort to minimize downstream effects, I've tried to preserve all of those for now.
* This temporarily includes a git dependency for `serde_spanned`, so that I can make the `miden-core-diagnostics` crate no-std compatible even with the `serde` feature enabled. I'm expecting a new release ASAP, but if one is not forthcoming in the next few days, I'll make the use of `serde_spanned` contingent on both `std` and `serde` features, and switch back to the crates.io release. This dependency is needed to make some of the source management types serializable with `serde`, which I rely on in subsequent changes I've got staged - it is introduced here for my convenience primarily.
* I've started moving dependency declarations used in multiple places into the root `Cargo.toml`, and using `dep.workspace = true` or `dep = { workspace = true, ... }` depending on the context. This drastically de-clutters the individual crate `Cargo.toml` files, and reduces the chance of us depending on divergent versions of the same dependencies. For now, I've focused only on a couple dependencies that are always needed by at least one crate in release builds; `dev-dependencies` and such I have not bothered with. I'd suggest we start to slowly move dependencies into the root `Cargo.toml` as we do further refactoring.

plafer commented on core-diagnostics/Cargo.toml:
```diff
@@ -0,0 +1,58 @@
+[package]
+name = "miden-core-diagnostics"
```
Please add this crate to `publish-release.sh` 😉

plafer commented on core-diagnostics/src/lib.rs:
```diff
@@ -0,0 +1,51 @@
+    related::{RelatedError, RelatedLabel},
+};
+
+pub(crate) mod utils {
+    pub use miden_crypto::utils::{
+        ByteReader, ByteWriter, Deserializable, DeserializationError, Serializable,
+    };
+}
```
What is the benefit of a pattern like this? Isn't it cleaner to, in this example, just use `miden_crypto::utils` in every file where it is used, as the reader can clearly see where these types are defined?

bobbinth replied:
> we're still a long ways away from being able to standardize our serialization around `serde`

This hasn't been a goal so far and actually we've tried to stay away from `serde` as much as possible. I'm fine with revisiting our serialization approach, especially since we'll probably need to move away from winterfell dependencies over the next few months - but we should probably figure out if `serde` is what we want to go with (this is a separate discussion though).

bitwalker replied:
Any particular reason to stay away from `serde`? IMO it is by far the most useful serialization library in the ecosystem, and nothing else even comes close. The biggest win is that we can trivially support serializing our types to different formats from a single description, which is of particular value with things like `Word`, `Felt`, `QualifiedProcedureName`, etc., that are used in many crates for different purposes.

plafer commented on core/Cargo.toml:
```diff
@@ -19,48 +19,27 @@ doctest = false
     "thiserror/std",
 ]
+serde = ["miden-core-diagnostics/serde", "miden-crypto/serde", "math/serde"]
+# DEPRECATED
+diagnostics = []
```
Let's remove this - every release comes with breaking changes anyways. Plus, I see that `miden-processor` still needlessly requires this feature (just noise at this point).

bobbinth replied:
I think removing this is fine, as @plafer mentioned, we are still introducing very significant breaking changes and I would rather introduce these now than later.

plafer commented on core/src/lib.rs:
```diff
@@ -50,14 +50,14 @@ assertion failed: `(left matches right)`
 
 mod kernel;
 pub use kernel::Kernel;
+pub use miden_core_diagnostics::debuginfo;
```
Did you add this to prevent another breaking change, or do you see us keeping this long term? 

I'm still wrapping my head around the best way to manage types between our crates; where I'm landed so far is that this wouldn't be desirable. That is, every `miden` type should be exported from a single crate; if someone wants the `debuginfo` module, they should import that crate. This is how [Plonky3](https://github.com/Plonky3/Plonky3/tree/main) does it, and I've found it quite pleasant to use, as I find it reduces the mental overhead of where to find things (since every type is only defined/exported in one place). You find yourself importing a bunch of `p3-*` crates, but that's fine IMO.

bitwalker replied:
> Did you add this to prevent another breaking change, or do you see us keeping this long term?

Purely for compatibility and avoiding more breaking changes, considering the laundry list of them at this point. Honestly, cleaning up the current disaster that is our legacy namespacing/re-exporting practices is going to be a long and/or painful process IMO.

> I'm still wrapping my head around the best way to manage types between our crates; where I'm landed so far is that this wouldn't be desirable. That is, every miden type should be exported from a single crate; if someone wants the debuginfo module, they should import that crate. 

Two things:

1. I absolutely 100% agree that we should not be re-exporting types from other crates _directly_ using e.g. `pub use external_crate::SomeType;` with very very few exceptions.
2. We _should_ re-export crates whose types appear in our public APIs, e.g. `pub use external_crate;`. This is to ensure that downstream users of that crate don't have to struggle with divergent dependency versions (which cause a caller's `SomeType` to differ from the callee's `SomeType`), they can simply import the type from the re-exported crate in their own code.

> This is how [Plonky3](https://github.com/Plonky3/Plonky3/tree/main) does it

I think this approach is really problematic for libraries as a general rule - it is really easy to get into a situation where your dependencies don't agree on a single version of a type. You pretty much should always re-export upstream crates for your users when types from those crates appear in your public API in order to avoid this.

bobbinth replied:
As mentioned in another comment, I think it is fine to get rid of this re-export.

plafer replied:
> I think this approach is really problematic for libraries as a general rule - it is really easy to get into a situation where your dependencies don't agree on a single version of a type.

And what is the problem with that? Say `crate-a` depends on `miden-core v0.15.0`, while `crate-b` depends on `miden-core v0.16.0`; then cargo will just include both versions in the final build. So other than (maybe) binary size, what is the problem/situation that we're trying to avoid here?

bitwalker replied:
@plafer The problem is when you want to pass a type from `miden-core v0.15.0` to an API that expects the `v0.16.0` version of that type, this is why you re-export the crate whose types are in your public API, so that one can always access the exact version of those types needed.

plafer commented on core/src/utils/mod.rs:
```diff
@@ -8,6 +6,7 @@ use core::{
 
 // RE-EXPORTS
 // ================================================================================================
+pub use miden_core_diagnostics::sync;
```
Similar question as in my previous comment

bobbinth replied:
My personal preference would be to keep `sync` in `miden-core` (at least for now). Places where we use it usually require `miden-core` anyway (though `miden-core-diagnostics` may be an exception here) and it would limit the amount of changes introduced in this PR.

This is a slight preference though - and if you feel strongly about it, I think it is fine to add another crate (though, we should probably re-organize directory structure of the repo as the number of crates grows - e.g., put them under `crates` directory as we do in `miden-base` and `miden-node`).

bitwalker replied:
I could go either way really - I'm just concerned about having things like a diagnostics crate take on a dependency to `miden-core` when it really is effectively standalone except for this one thing (the `sync` module). One of the problems we have at the moment (IMO) is the matter of where to place types in such a way that we can use them in all the places they are needed, without ending up in a situation where we can't depend on something somewhere due to circular dependencies.

Consider that `Library`, `LibraryPath`, etc., were defined in `miden-assembly`, then split out into `miden-assembly-syntax`, even though libraries aren't really a thing that that crate needs to know about, aside from paths as the means of qualifying module/procedure names. We would like that stuff to be the purview of `miden-mast-package`, but that crate depended on `miden-assembly` so that it could assemble packages in its tests, which mean that we couldn't move the types to `miden-mast-package` without ending up with a circular dependency (or having duplicate definitions of things like `LibraryPath`). Basically, the more entangled our dependency graph, the more frequently we're going to either end up with awkward placement of code in order to satisfy the needs of the graph, or needless duplication of types for the same reason. The narrower the scope of individual crates, the less entanglement of this kind we have, and we gain other benefits as well (better incremental compilation, faster compilation in general, more reuse, etc.). The downside is that it makes our already messy namespacing issue at the crate level, more so, at least for now.

Anyway, I think for now I'll leave it in `miden-core`, but just wanted to explain in more detail why more generally we probably should prefer splitting standalone things into standalone crates, rather than having chonky all-in-one crates.

plafer reviewed (approved):
LGTM!

We should have a conversation about how exactly we want to manage crates/exports/reexports, as alluded to in this PR description, some of my comments, offline conversations, etc. As mentioned in my comments, I generally like how Plonky3 does it, but would be curious if there are any downsides to that approach.

bobbinth commented on core/Cargo.toml:
```diff
@@ -19,48 +19,27 @@ doctest = false
-thiserror = { workspace = true }
-winter-utils = { package = "winter-utils", version = "0.13", default-features = false }
+thiserror.workspace = true
+winter-utils = { version = "0.13", default-features = false }
```
We could almost remove this dependency:

- It is primarily used to import serialization-related types, which could come from `miden-crypto` instead.
- One serialization-related type that isn't re-exported from `miden-crypto` is `ReadAdapter` - but it probably should be.
- The only other missing thing would be `group_slice_elements()` which as of Rust 1.88 could be replaced with `slice::as_chunks()`.

bitwalker replied:
I should also note that I think we can remove the serialization of `.masl` files - there isn't any reason to support that anymore IMO, and it is effectively just an incomplete subset of what gets serialized to a `.masp` file anyway.

That would remove the need for `ReadAdapter` in `miden-assembly-syntax` (and indeed the need for it to depend on any of the `winter-utils`/`miden-crypto` serialization bits).

bobbinth replied:
> I think we should bump the MSRV to latest stable, we'll pick up quite a few nice improvements, and there isn't really a compelling reason to hold our MSRV back on older releases at the moment. Especially if it enables further clean up of our deps.

Agreed! But maybe do it in another PR?

> I'd like to revisit whether we actually need the `winter-utils` serialization primitives, or if we can replace them with `serde`.

Agreed - I wrote in a bit more detail about it in https://github.com/0xMiden/miden-vm/pull/1945#discussion_r2188545005.

> Since the only real reason we have any serialization-related code in `miden-core` in the first place is for the serialization of MAST, which should probably be the responsibility of the Miden package format

Agreed - though I'm not sure `serde`-based serialization will really work for MAST. One reason is that we actually will need to serialize it in 2 different ways:

- Compact but slow to deserialize - in this format we won't be serializing internal node hashes and would derive them from the basic blocks + some metadata.
- Large but fast to deserialize - this would work largely as it works now, I estimate that this format would result in 5x - 10x bigger serialized files than the compact one.

But maybe there is a way to make it work with `serde` as well.

bitwalker replied:
> Agreed! But maybe do it in another PR?

For sure 👍 

> * Compact but slow to deserialize - in this format we won't be serializing internal node hashes and would derive them from the basic blocks + some metadata.
>
> * Large but fast to deserialize - this would work largely as it works now, I estimate that this format would result in 5x - 10x bigger serialized files than the compact one.

Certainly both can be done with `serde`, but supporting both simultaneously might be more of a pain, I'd need to consider it more to see if there is an elegant way to do so with `serde` out of the box, or whether it would be easier to support the fast format with `serde` and the slow format with something hand-rolled (that may re-use bits of the `serde` serialization where there is overlap).

Regardless, we only have the fast format currently, so it would make sense to use `serde` for that now and then make the decision on whether to implement the slow format with `serde` as well, or hand-roll that one, when we get to the point of implementing it, IMO.

bobbinth commented on Cargo.toml:
```diff
@@ -37,4 +38,22 @@ debug-assertions = true
 overflow-checks = true
 
 [workspace.dependencies]
```
In `miden-base` and other crates we've started to group dependencies by their categories (e.g., see [here](https://github.com/0xMiden/miden-base/blob/next/Cargo.toml#L40). Maybe makes sense to do this here as well?

bitwalker replied:
Sure!

bobbinth commented on assembly-syntax/Cargo.toml:
```diff
@@ -32,10 +31,8 @@ aho-corasick = { version = "1.1", default-features = false }
 env_logger = { version = "0.11", optional = true }
 lalrpop-util = { version = "0.22", default-features = false }
 log = "0.4"
```
nit: `log` could now be `log.workspace = true`.

bobbinth commented on core-diagnostics/Cargo.toml:
```diff
@@ -0,0 +1,58 @@
+parking_lot = { version = "0.12", optional = true }
+paste = "1.0"
+serde = { workspace = true, optional = true }
+serde_spanned = { version = "0.6", optional = true, default-features = false }
```
The PR description mentions that `serde_spanned` is a git dependency, but seems like we are able to rely on the crates.io version here?

bitwalker replied:
Yes, I went ahead and switched to the crates.io version for now - the new release is coming this week I believe, but to avoid holding up this PR, I made some adjustments. I have another PR pending where I can do the upgrade to that release when it happens.

bobbinth reviewed (commented):
Looks good! Thank you! (and really appreciate some of the dependency-related clean-up). I've replied to some comments above and also left a few more comments/questions below.

bitwalker commented:
@bobbinth @plafer A few updates so you know what has changed since your review:

**NOTE:** All of the changes described below are made in separate commits so they are more easily reviewed independently.

* I split up the diagnostics and source-level debug information components into two crates, `miden-utils-diagnostics` and `miden-debuginfo` (I'll get to naming in a minute).
* I did end up splitting out the `sync` module into a new crate, `miden-utils-sync`. This was necessary to disentangle awkward dependencies between `miden-core`, `miden-utils-diagnostics`, and `miden-debuginfo`. Namely, `miden-core` requires the `miden-debuginfo` types `Location` and `Uri`, while `miden-debuginfo` requires the `sync` module for the `DefaultSourceManager` implementation, which we wanted to provide from `miden-core`. Rather than try to split up the `miden-debuginfo` types across `miden-core` and `miden-debuginfo`, it is more straightforward to extract the `sync` module to a new crate, make `miden-debuginfo` and `miden-core` depend on it, and then `miden-core` can depend on `miden-debuginfo` for access to `Location` and `Uri`.
* In the process of doing the above, I created a `crates` directory under which we can start to move crates, starting with the new ones I've added. This directory is meant to be organized to reflect the naming, e.g. `miden-utils-diagnostics` is in `crates/utils/diagnostics`, while `miden-debuginfo` is in `crates/debuginfo`. This is so we can start cleaning up the project root a bit with all of the crates we have.
* It also felt appropriate here to move `miden-test-utils` to `crates/utils/testing` and set up the workspace-level dependency so that it appears as `miden-utils-testing` (so the crate naming is consistent), while preserving the actual crate name of `miden-test-utils` for now, so we can continue to publish it under that name. We don't appear to be automatically publishing that crate though, and it isn't clear to me why we would be, other than perhaps to satisfy Cargo publishing checks - but regardless, my intent was to try and canonicalize the naming here around `miden-utils-<crate>` (for general utility crates).
* As a result of all this, I also undertook to make our workspace `Cargo.toml` files consistent about using workspace dependencies, the actual crate names rather than crate aliases, and simplify re-exports from `miden-core`, `miden-assembly-syntax`, and `miden-assembly`.
* I've also incorporated the other review feedback I hadn't yet addressed

The above changes have resulted in a fair amount of churn, mostly trivial renames, but hopefully this is a one-time annoyance and things will be easier from here on.

I'd suggest we merge this as soon as possible to avoid conflicts with other ongoing work, but it's pretty unavoidable that there will be conflicts. I'm happy to also resolve the conflicts myself, here, once all other work going into the `0.16.0` release is ready/merged, and only merge this PR then - let me know one way or the other.

bitwalker commented on miden-vm/benches/program_execution_fast.rs:
```diff
@@ -54,7 +54,7 @@ fn program_execution_fast(c: &mut Criterion) {
                     bench.to_async(Runtime::new().unwrap()).iter_batched(
                         || {
-                            let processor = FastProcessor::new_with_advice_inputs(
+                            let processor = Fastmiden_processor::new_with_advice_inputs(
```
Whoops, good catch

bitwalker commented on miden-vm/tests/integration/cli/cli_test.rs:
```diff
@@ -63,8 +63,8 @@ fn cli_run() -> Result<(), Box<dyn std::error::Error>> {
+use miden_core::Decorator;
 
 #[test]
 fn cli_bundle_debug() {
```
Thanks, I'll have a look!

bitwalker replied:
@plafer I'm unable to reproduce with `make test` locally - and it looks like the tests are being run. Which test is hanging for you?

bobbinth replied:
I've restarted the task in the CI but it seems like it is going to hang again. The tests that seem to cause this are:

- `cli_bundle_debug`.
- `cli_bundle_kernel`.
- `cli_bundle_kernel_noexports`.

bitwalker replied:
@plafer What system are you testing on, macOS or Linux? If the latter, what arch?

bitwalker replied:
@bobbinth @plafer ^

bobbinth replied:
Would the solution be to disable CLI tests for now? We have just a few of them (10, I think?) and while important, they don't test anything super critical.

plafer replied:
Thank you for investigating! I think we should indeed disable them to not hold up this PR, and create an issue with the suggested fix (i.e. making them unit tests)

bitwalker replied:
I've found the actual bug, but also modified our test configuration to run the CLI tests serially (so we only run `cargo build` for `miden-vm` once), and to enforce a timeout so that tests don't hang our CI runners for hours.

plafer replied:
Interesting, so what was it about the environment variables (and `internal` feature?) that caused the bug? (based on [128f4ba](https://github.com/0xMiden/miden-vm/pull/1945/commits/128f4ba491a4ed3b0b30bfe8e40a908d43e3ed29))

bitwalker replied:
* For the environment variables, `escargot` uses `--message-format=json` under the covers, but we were forcing Cargo to emit more verbose (and colored) output, which resulted in the parsing of messages by `escargot` to fail, and internally that was causing it to hang. The environment variables we were setting don't make sense to set here anyway, because not only do we not see the output, the output is already verbose because of the message format used.
* The `internal` feature is implied by `executable`, but furthermore, there appears to be an issue with how `escargot` is passing the features through to `cargo`. They are supposed to be comma-separated, not space-separated, and from what I can tell, `escargot` is passing them through unmodified. I can't tell if this was a change made in Cargo, or what. Regardless, we didn't need to specify the `internal` feature anyway.

bobbinth commented on crates/debuginfo/Cargo.toml:
```diff
@@ -0,0 +1,38 @@
+[package]
+name = "miden-debuginfo"
```
nit (feel free to ignore): would `miden-debug-types` be a more appropriate name?

bitwalker replied:
That seems reasonable to me, `miden-debuginfo` was all I could come up with that seemed reasonable, but I like `miden-debug-types`

bobbinth commented on crates/utils/diagnostics/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+[package]
+name = "miden-utils-diagnostics"
```
Question: the main reasons why this is `miden-utils-diagnostics` rather than `miden-diagnostics` is that `miden-diagnostics` is already taken by our other crate, right?

But if this is VM-specific, should it be `miden-vm-diagnostics`?

bitwalker replied:
Correct - I would've used `miden-diagnostics` if it wasn't currently taken, and I would've moved this stuff to `miden-diagnostics` if it was relatively easy to do so. Unfortunately, the AirScript crates depend on the current implementation of `miden-diagnostics`, and migrating is a bit more involved than I wanted to tackle as part of this (and would've ended up imposing another dependency on releasing these changes).

This crate isn't really VM-specific though, or really even VM-related (to my knowledge, the VM doesn't really use it to nearly the same extent as the compiler and assembler). Ultimately the end goal would be to move this to `miden-diagnostics` in a future release, once we can migrate AirScript, thereby unifying most of our toolchain components around a common diagnostics infrastructure. So the naming here is really temporary IMO.

bobbinth commented on crates/utils/diagnostics/README.md:
Not for this PR, but would be good to add a bit more info to the README (we do this for most other crates). Let's create an issue for this.

bobbinth commented on crates/utils/sync/README.md:
Same comment about the readme file as above - ideally, we'd provide a bit more here to be consistent with our other crates.

bobbinth replied:
I don't think we need to add more than a couple of paragraphs. Maybe list the what structs are provided and some basic info about implementation (e.g., that we use spin-lock). I would also add license section at the bottom as we do in other crates (e.g., how we do in the [package](https://github.com/0xMiden/miden-vm/tree/next/package) crate).

But again, I think we can do it in a separate PR - so, I'll merge this one as is.

bobbinth reviewed (approved):
Looks good! Thank you! I left a few comments/questions inline. They are pretty much optional - so, feel free to merge even as is.
